### PR TITLE
Serial

### PR DESCRIPTION
**Pull request by caitlin-hennessy**
This is an _in-progress_ PR - see my email for details
serial_editor_2

### CODE REVIEW

File: pxtsim/logs.ts [hunk 1]
```diff
@@ -190,50 +199,51 @@ namespace pxsim.logs {
...
-                                    break;
-                                default:
-                                    buffer += value[i];
-                                    if (buffer.length > (this.props.maxLineLength || 255)) {
-                                        this.appendEntry(source, buffer, theme);
-                                        buffer = ''
-                                    }
-                                    break;
+                        const sim = !!smsg.sim || false;
+                        if (sim == this.isSim) {
```
Comment (pelikhan) [inline pxtsim/logs.ts]: break instead of nesting

File: pxtsim/logs.ts [hunk 2]
```diff
@@ -190,50 +199,51 @@ namespace pxsim.logs {
...
-                                case 13: //'\r'
-                                    break;
-                                default:
-                                    buffer += value[i];
-                                    if (buffer.length > (this.props.maxLineLength || 255)) {
-                                        this.appendEntry(source, buffer, theme);
-                                        buffer = ''
-                                    }
-                                    break;
+                        const sim = !!smsg.sim || false;
```
Comment (pelikhan) [inline pxtsim/logs.ts]: You can drop ``|| false``

File: pxtsim/logs.ts [hunk 3]
```diff
@@ -290,6 +300,18 @@ namespace pxsim.logs {
                 e.element.className = "ui log " + e.theme;
                 let raiseTrends = false;
                 if (e.accvalues) {
+                    e.accvaluesElement = document.createElement("span");
+                    e.accvaluesElement.className = "ui log " + e.theme + " gauge"
+                    e.chartElement = document.createElement("canvas");
+                    new CanvasChart().drawChart(e.chartElement, [new Point(e.accvalues[0].t, e.accvalues[0].v)])
```
Comment (pelikhan) [inline pxtsim/logs.ts]: drawing chart should be deferred

Comment (pelikhan) [reply pxtsim/logs.ts]: We should make the CanvasChart work with the .t, .v values instead of allocation "new Point" objects.

File: pxtsim/logs.ts [hunk 4]
```diff
@@ -335,14 +360,15 @@ namespace pxsim.logs {
 
                 if (entry.countElement) entry.countElement.innerText = entry.count.toString();
                 if (entry.accvaluesElement) entry.accvaluesElement.innerText = entry.value;
-                if (entry.chartElement) entry.chartElement.render();
+//                if (entry.chartElement) entry.chartElement.render();
+                if (entry.chartElement) new CanvasChart().drawChart(entry.chartElement, entry.accvalues.map(accvalue => new Point(accvalue.t, accvalue.v)))
```
Comment (pelikhan) [inline pxtsim/logs.ts]: update canvaschart to take accvalues as input instead of allocating new point objects.

File: webapp/src/app.tsx [hunk 5]
```diff
@@ -862,6 +864,8 @@ export class ProjectView
         if (options.filesOverride)
             Util.jsonCopyFrom(files, options.filesOverride)
         files["pxt.json"] = JSON.stringify(cfg, null, 4) + "\n"
+        //TODO this
+        files["serial.json"] = JSON.stringify({"pineapple": "banana"}) + "\n"
```
Comment (pelikhan) [inline webapp/src/app.tsx]: serial.txt?

File: pxtsim/canvaschart.ts [hunk 6]
```diff
@@ -0,0 +1,344 @@
+
```
Comment (pelikhan) [inline pxtsim/canvaschart.ts]: is this file needed to go away?

File: webapp/src/logview.tsx [hunk 7]
```diff
@@ -37,6 +44,8 @@ export class LogView extends React.Component<{}, LogViewState> {
             nameFilter: serial.nameFilter
         })
         this.state = {};
+
+        this.view.setLabel(this.props.isSim ? lf("SIM") : ("DEV"));
```
Comment (pelikhan) [inline webapp/src/logview.tsx]: Don't use acronyms

File: webapp/src/serial.tsx [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    private maxLineLength: number = 500
+    private maxAccValues: number = 500
+    private renderFiberId: number
+    private maxConsoleEntries: number = 20
+
+    private active: boolean = true
+
+    acceptsFile(file: pkg.File) {
+        // TODO hardcoded string
+        return file.name === "serial.json"
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: define the constant here: https://github.com/Microsoft/pxt/blob/master/pxtlib/main.ts#L211

File: webapp/src/serial.tsx [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+        super(parent)
+        window.addEventListener("message", this.processMessage.bind(this), false)
+    }
+
+    processMessage(ev: MessageEvent) {
+        let msg = ev.data
+        if (this.active && msg.type === "serial") {
+            const smsg = msg as pxsim.SimulatorSerialMessage
+            const sim = !!smsg.sim || false
+            if (sim == this.isSim) {
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: break instead of nesting

File: webapp/src/serial.tsx [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+    scheduleRender() {
+        if (!this.renderFiberId) this.renderFiberId = setTimeout(() => this.render(), 50);
+    }
+
+    render() {
+        this.consoleEntries.forEach(e => {
+            if (!e.dirty) return
+            let consoleDiv = document.createElement("div")
+            consoleDiv.className = "coconut"
```
Comment (pelikhan) [inline webapp/src/serial.tsx]: coconut?

**Comment by microsoftsam** [comment]
Could you add some screenshots if there are UI changes included?

File: pxtsim/logs.ts (hunk 2 above)
Comment (caitlin-hennessy) [reply pxtsim/logs.ts]: Fixed

File: pxtsim/logs.ts (hunk 1 above)
Comment (caitlin-hennessy) [reply pxtsim/logs.ts]: Any particular reason why?  This is a bit more concise, and should compile to the same thing.

File: pxtsim/logs.ts (hunk 3 above)
Comment (caitlin-hennessy) [reply pxtsim/logs.ts]: Drawing chart actually _is_ deferred in serial.tsx, which is where I'm porting the functions from logs.ts.  My bad - logs.ts wasn't supposed to be in this PR.

File: pxtsim/logs.ts (hunk 1 above)
Comment (pelikhan) [reply pxtsim/logs.ts]: Personal preference to avoid nesting too much. Disregard.

Comment (caitlin-hennessy) [reply pxtsim/logs.ts]: ok

File: webapp/src/app.tsx (hunk 5 above)
Comment (caitlin-hennessy) [reply webapp/src/app.tsx]: sure - the file is for all practical purposes empty, so can have whatever extension we want.

File: pxtsim/canvaschart.ts (hunk 6 above)
Comment (caitlin-hennessy) [reply pxtsim/canvaschart.ts]: yes - done.

File: webapp/src/serial.tsx (hunk 8 above)
Comment (caitlin-hennessy) [reply webapp/src/serial.tsx]: Done

File: webapp/src/logview.tsx (hunk 7 above)
Comment (caitlin-hennessy) [reply webapp/src/logview.tsx]: Fixed

File: pxtsim/logs.ts (hunk 4 above)
Comment (caitlin-hennessy) [reply pxtsim/logs.ts]: Done

File: webapp/src/serial.tsx (hunk 10 above)
Comment (caitlin-hennessy) [reply webapp/src/serial.tsx]: oops...fixed!
